### ISSUE TITLE
c.DockerSpawner.volumes invalid characters for a local volume 

### ISSUE THREAD

**Issue by yocode** [jupyterhub/dockerspawner#209]
This is related to how {username} gets defined when an OAuth user comes in to create a docker volume.

My current config:
`notebook_dir` = '/home/jovyan/work'
c.DockerSpawner.notebook_dir = notebook_dir
c.DockerSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir `}`

The resultant error when a server is started:

 400 Client Error: Bad Request ("create jupyterhub-user-X.Y.Z@gmail.com: "jupyterhub-user-X.Y.Z@gmail.com" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]"

I though i might be able to populate the c.DockerSpawner.volumes dictionary with regex to match everything before '@', but so far been unsuccessful because {username} is populated at runtime and not at jupyterhub init

c.DockerSpawner.volumes = { re.search(r'.*(?=@)', '{username}') : notebook_dir }

Any thoughts as to how I can accomplish this?

**Comment by dan-gittik** [comment]
I had a similar issue, and found out that according to [this code](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/google.py#L89), `@...` is actually stripped - _if_ you define the [`HOSTED_DOMAIN`](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/google.py#L44) environment variable.
So, adding `HOSTED_DOMAIN=gmail.com` in `oauth.env` or `c.GoogleOAuthenticator.hosted_domain = 'gmail.com'` in `jupyterhub_config.py` ([see here](https://github.com/jupyterhub/oauthenticator#google-setup)) should do it.

**Comment by bellackn** [comment]
I would like to jump right in here. I face the same problem, having the same config as @yocode.
```
[E 2019-03-25 08:53:27.381 JupyterHub user:477] Unhandled error starting xxx@xxx.de's server: 400 Client Error: Bad Request ("create jupyterhub-xxx@xxx.de: "jupyterhub-xxx.xxx@xxx.de" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path")
```
The users in my case come from a Keycloak instance and authenticate via OIDC. I don't see where I could use something similar to the `HOSTED_DOMAIN` suggestion by @dan-gittik, especially because I don't know the domains beforehand, because users can use self-registration there.

I'm willing to contribute something myself if someone can give me a hint where to look. I suspect it somewhere here, but I don't understand enough of Python to get any further: https://github.com/jupyterhub/jupyterhub/blob/8032f874affd541764d2a34d481a2bb4f7d9ceb7/jupyterhub/spawner.py#L766

**Comment by kczekirda** [comment]
@bellackn please try to update your  dockerspawner, it should we fixed now.

**Comment by bellackn** [comment]
Hi @kczekirda, thanks for your response. Unfortunately, the new version comes up with a new error, which is described in #290. 
As pointed out there, commenting `c.DockerSpawner.notebook_dir = notebook_dir` works and the container starts properly – even when there are special characters like "@" in the usernames! But I don't fully understand the consequences of commenting out said line, so I guess I will stick with Dockerspawner version 0.10 for now, as this seems like a cleaner solution to me.

**Comment by kkr78** [comment]
notebook_dir = os.environ.get('DOCKER_NOTEBOOK_DIR') or '/home/jovyan/work'
c.DockerSpawner.notebook_dir = notebook_dir
c.DockerSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir }

We have similar issue. Azure AD returns user name with comma and space ("lastname, firstname"). How can I resolve this error? We don't want users to see each other files. are there any consequences if I comment the last line? 

[E 2019-07-05 17:23:46.007 JupyterHub user:626] Unhandled error starting xx, yy server: 400 Client Error: Bad Request ("create jupyterhub-user-xx, yy: "jupyterhub-user-xx, yy" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path")

**Comment by anmtan** [comment]
same issue as @kkr78. Any suggestions?

**Comment by kkr78** [comment]
> same issue as @kkr78. Any suggestions?

@anmtan try this

from oauthenticator.azuread import LocalAzureAdOAuthenticator
class NormalizedUsernameLocalAzureAdOAuthenticator(LocalAzureAdOAuthenticator):
    def normalize_username(self, username):
        return username.lower().replace(',', '').replace(' ','_') ##replace all special characters here

c.JupyterHub.authenticator_class = NormalizedUsernameLocalAzureAdOAuthenticator

**Comment by anmtan** [comment]
Thanks @kkr78. I tried with your suggestion. It failed with user not found. 

Let me clarify my question. My goal is to use AzureAdOAuthenticator, and be able to mount container to host volume in order to persist data. I had exactly the same error in your earlier post because Azure AD returns 'lastname, firstname' as the username. I upgraded dockerspawner==0.11.1. (I was using 0.10.0) After upgrading dockerspawner, that error is gone. See logs below.

JupyterHub base:499] User logged in: lastname, firstname
...
JupyterHub user:190] Creating <class 'dockerspawner.dockerspawner.DockerSpawner'> for lastname, firstname:
...
JupyterHub dockerspawner:873] Starting host with config: {'binds': {'jupyterhub-user-lastname_2C_20firstname': {'bind': '/home/jovyan/work', 'mode': 'rw'}}
JupyterHub dockerspawner:990] Created container jupyter-lastname_2C_20firstname


This is what I use in jupyterhub_config.py


from oauthenticator.azuread import AzureAdOAuthenticator
c.JupyterHub.authenticator_class = AzureAdOAuthenticator
c.Application.log_level = 'DEBUG'
c.AzureAdOAuthenticator.tenant_id = os.environ.get('AAD_TENANT_ID')
c.AzureAdOAuthenticator.oauth_callback_url = 'http://{your-domain}/hub/oauth_callback'
c.AzureAdOAuthenticator.client_id = '{AAD-APP-CLIENT-ID}'
c.AzureAdOAuthenticator.client_secret = '{AAD-APP-CLIENT-SECRET}'

notebook_dir = os.environ.get('DOCKER_NOTEBOOK_DIR') or '/home/jovyan/work'
c.DockerSpawner.notebook_dir = notebook_dir

# Mount the real user's Docker volume on the host to the notebook user's
# notebook directory in the container
c.DockerSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir }


The problem that I am facing is after user logged in, the container was also created. But I am not able to create/save any notebook. And I am suspecting because of the username mismatch. Here is the log.

[I 2019-08-03 15:16:57.771 SingleUserNotebookApp handlers:153] Creating new notebook in /
[I 2019-08-03 15:16:57.807 SingleUserNotebookApp sign:396] Writing notebook-signing key to /home/notebook/.local/share/jupyter/notebook_secret
[W 2019-08-03 15:16:57.827 SingleUserNotebookApp web:1667] 403 POST /user/lastname%2C%20firstname/api/contents/?1564845417531 (::ffff:10.7.2.173): Permission denied: Untitled.ipynb
[W 2019-08-03 15:16:57.827 SingleUserNotebookApp handlers:568] Permission denied: Untitled.ipynb
[W 2019-08-03 15:16:57.828 SingleUserNotebookApp log:158] 403 POST /user/lastname%2C%20firstname/api/contents/?1564845417531 (lastname, firstname@::ffff:10.7.2.173) 57.47ms
[I 2019-08-03 15:16:58.852 SingleUserNotebookApp log:158] 200 GET /user/lastname%2C%20firstname/api/sessions?1564845418611 (lastname, firstname@::ffff:10.7.2.173) 1.12ms


In the fixes of 0.11.1 by @minrk, the username was normalized to "jupyterhub-user-lastname_2C_20firstname" but the user in the store still "/user/lastname%2C%20firstname", I am not sure if that causes the Permission denied. I am just suspecting the username mismatch in my case. Any thoughts?

**Comment by kkr78** [comment]
I completely forgot LocalAzureAdOAuthenticator only worked on default spwaner. please ignore that.

But with Docker spawner, I am able to save the notebooks without any issue on user docker container after commented the below line. otherwise, I have seen the similar error reported above "includes invalid characters for a local volume name". were u able to make it work?

##c.DockerSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir }

which docker image did you configure for docker spawner? I configured jupyter/pyspark-notebook:latest

Did u try to login to user docker container and check the permissions? 

jovyan@2a473d21a05b:~$ pwd
/home/jovyan
jovyan@2a473d21a05b:~$ ls -ltr
total 8
drwsrwsr-x 2 jovyan users 4096 Apr  8 23:37 work
-rw-r--r-- 1 jovyan users 1252 Aug  5 03:02 Test.ipynb
jovyan@2a473d21a05b:~$

**Comment by anmtan** [comment]
This line mount the docker container to the host. I think you should have this to bind the host dir to container dir. I tried comment it out, it just failed to start the container.

c.DockerSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir }

I also checked the permission in the container, everything looks fine. When we use DummyAuthenticator, with username without comma or spaces, everything works fine. 

I think the problem is on the URL encoded username. See docker container logs below

[W 2019-08-05 16:35:41.724 SingleUserNotebookApp web:1667] 400 PUT /user/lastname%2C%20firstname/lab/api/workspaces/user/lastname%2C%20firstname/lab?1565022941464 (::ffff:10.7.2.173): Workspace metadata ID mismatch: expected 'user/lastname, firstname/lab' got '/user/lastname%2C%20firstname/lab'
[W 2019-08-05 16:35:41.725 SingleUserNotebookApp handlers:568] Workspace metadata ID mismatch: expected 'user/lastname, firstname/lab' got '/user/lastname%%2C%%20firstname/lab'

**Comment by ryan-daly-AgFCS** [comment]
Has there been any solutions to the AAD %2C%20 clean up above? We're experiencing this right now as we build out our containers, and would like to not have to fix folder issues on the network once we do roll it out.

**Comment by minrk** [comment]
This will be closed by #414

**Comment by bakcsa83** [comment]
I'm also experiencing this issue.
In my jupyterhub_config I have the followings:
```
import os
from oauthenticator.generic import GenericOAuthenticator
c.JupyterHub.authenticator_class = GenericOAuthenticator

#....
c.DockerSpawner.volumes = { '/user-data/jupyter-home/{username}': notebook_dir }
#....
```
It works fine with usernames that do not have special characters.

However, if `john_doe` logs in, then the spawner tries to mount `/user-data/jupyter-home/john-5fdoe` which is obviously wrong.
